### ISSUE TITLE
Design suggestion for better type check of has-aux

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jaxopt#332]
Currently, various optimizers accept a generic `Callable` function and parameters for specifying how it should be used:
```python
class LBFGS:
  fun: Callable
  value_and_grad: bool = False
  has_aux: bool = False
```
So, the user has to carefully initialize making sure that these three variables are consistent.

Instead, it might be a better user experience to have:
```python
RealNumeric = jax.Array | float

class _ValueAuxAndGradProtocol(Protocol, Generic[T]):
    def __call__(self, x: T, /, *args: Any, **kwargs: Any
                 ) -> tuple[tuple[RealNumeric, Any], T]:
      ...

class _HasAuxProtocol(Protocol, Generic[T]):
    def __call__(self, x: T, /, *args: Any, **kwargs: Any
                 ) -> tuple[RealNumeric, Any]:
      ...

class ValueAuxAndGrad(Generic[T]):
  fun: _ValueAuxAndGradProtocol

  def __call__(self, x: T, /) -> tuple[tuple[RealNumeric, Any], T]:
      return self.fun(x, *args, **kwargs)

class HasAux(Generic[T]):
  fun: _HasAuxProtocol

  def __call__(self, x: T, /, *args: Any, **kwargs: Any
               ) -> tuple[RealNumeric, Any]:
      return self.fun(x, *args, **kwargs)

class LBFGS(Generic[T]):
  fun: Callable[[T], RealNumeric] | ValueAuxAndGrad[T] | HasAux[T]
  # No need for flags!
```
This way, the function parmeters and return values are type-checked in all three cases.  Also, there's no need for any flags.  Instead of
```python
LBFGS(f, has_aux=True, value_and_grad=True)
```
you can do
```python
LBFGS(ValueAuxAndGrad(f))  # type-checked
```
What do you think?

**Comment by NeilGirdhar** [comment]
(Just discovered this library by the way, and it's a real gem!!)

**Comment by mblondel** [comment]
Thanks a lot, this is an interesting suggestion. 

One advantage of `has_aux` and `value_and_grad` is that those are familiar keywords for JAX users. I'm a bit concerned that your approach would make JAXopt too "frameworkish".

But we are aware that we should improve error checking, i.e., make sure that `fun`'s output is compatible with what has been specified in the `has_aux` and `value_and_grad` options.

**Comment by NeilGirdhar** [comment]
@mblondel That's fair enough, but the benefit to doing this in Jax is smaller since it's just one flag whereas here you hav two flags, and a somewhat complicated interacting that requires reading the documentation carefully when using them (if you only use one of the flags, which one are you allowed to use?—it's not obvious without reading the docs).  So that's why I proposed this since instead of reading the documentation to understand what code does, it becomes self-documenting.

Also, since you just generalized the has-aux/value-and-grad to all the optimizers, this woudl be a good time to make any architectural changes before people start using the flags.

**Comment by mblondel** [comment]
I'm not saying your idea is bad but I would like to understand its pros and cons and I would like JAXopt to integrate well in the JAX ecosystem...

If we adopt objects to represent functions, this could also potentially be useful so that functions know how to initialize the parameters.

**Comment by mblondel** [comment]
Thanks for starting the discussion in google/jax#12948. This would have the advantage of being able to use `jax.*` operators instead of creating new ones in JAXopt.

**Comment by NeilGirdhar** [comment]
@mblondel Yeah, I started the discussion, but I wasn't sure if I should leave it open since the benefits are smaller in Jax (as I explained in my last comment).  I'll reopen and maybe someone from the Jax team will comment though.  There may be benefits that I'm missing.

**Comment by mblondel** [comment]
This would be useful to "annotate" the returned Callable with its structure.
